File: taurus-protect-sdk-typescript/tests/unit/helpers/signed-fixtures.test.ts
/**
 * Cross-SDK signed fixtures for the two verification thresholds.
 *
 * Step 2 is the tenant-wide SuperAdmin threshold; step 5 is the per-group one. Different
 * scopes, same counting rule, so one file gates both.
 *
 * The behaviour vectors cover everything expressible without key material; this file
 * covers what needs real signatures. Until it existed, the repo's own CLAUDE.md
 * recorded that NO automated cross-SDK gate covered the distinct-key rule, and the
 * five cases below lived as five hand-maintained copies in four suites.
 *
 * The fixture carries PUBLIC keys and signatures only — the gate verifies, it never
 * signs — so there is no private key material in the repo.
 */

import * as crypto from "crypto";
import { readFileSync, existsSync } from "fs";
import * as path from "path";

import { decodePublicKeyPem, encodePublicKeyPem } from "../../../src/crypto";
import { ConfigurationError, IntegrityError } from "../../../src/errors";
import { verifyGovernanceRulesSignatures } from "../../../src/helpers/signature-verifier";
import { WhitelistedAddressVerifier } from "../../../src/helpers/whitelisted-address-verifier";
import type {
  DecodedRulesContainer,
  GroupThreshold,
} from "../../../src/models/governance-rules";
import type { WhitelistSignatureEntry } from "../../../src/models/whitelisted-address";

// tests/unit/helpers -> tests/unit -> tests -> <sdk> -> <repo root>
const FIXTURES_PATH = path.resolve(
  __dirname,
  "../../../../scripts/resources/verification-signed-fixtures.json"
);

interface ThresholdCase {
  description: string;
  signatures: Array<{ user_id: string; signature: string }>;
  super_admin_keys_pem: string[];
  min_valid_signatures: number;
  expect: "ok" | "error";
}

interface GroupCase {
  description: string;
  metadata_hash: string;
  group_id: string;
  minimum_signatures: number;
  users: Array<{ user_id: string; public_key_pem: string }>;
  group_user_ids: string[];
  signatures: Array<{ user_id: string; signature: string; hashes: string[] }>;
  expect: "ok" | "error";
}

interface Fixtures {
  rules_container_base64: string;
  counts: Record<string, number>;
  superadmin_threshold: ThresholdCase[];
  group_threshold: GroupCase[];
}

/** Loaded inside `it()`, never in a describe body — a throw during collection drops
 * the test count instead of showing a red test. */
function loadFixtures(): Fixtures {
  if (!existsSync(FIXTURES_PATH)) {
    throw new Error(`cannot read shared signed fixtures ${FIXTURES_PATH}`);
  }
  const fixtures = JSON.parse(readFileSync(FIXTURES_PATH, "utf-8")) as Fixtures;

  for (const [key, expected] of Object.entries(fixtures.counts)) {
    const actual = (fixtures as unknown as Record<string, unknown[]>)[key].length;
    expect([key, actual]).toEqual([key, expected]);
  }
  expect(fixtures.rules_container_base64).toBeTruthy();
  return fixtures;
}

describe("signed fixtures: SuperAdmin threshold", () => {
  it("agrees with every recorded outcome", () => {
    const fixtures = loadFixtures();
    const container = Buffer.from(fixtures.rules_container_base64, "base64");

    for (const c of fixtures.superadmin_threshold) {
      const keys = c.super_admin_keys_pem.map(decodePublicKeyPem);
      const signatures = c.signatures.map((s) => ({ signature: s.signature }));

      let threw: unknown;
      try {
        verifyGovernanceRulesSignatures(
          container,
          signatures,
          keys,
          c.min_valid_signatures
        );
      } catch (error: unknown) {
        threw = error;
      }

      if (c.expect === "error") {
        // Assert as a tuple so a failing loop iteration names itself.
        expect([c.description, threw instanceof Error]).toEqual([c.description, true]);
        expect(
          threw instanceof IntegrityError || threw instanceof ConfigurationError
        ).toBe(true);
      } else {
        expect([c.description, threw]).toEqual([c.description, undefined]);
      }
    }
  });

  // Without a case where the entry COUNT meets the threshold but the distinct-key
  // count does not, the whole file would pass against an implementation that counts
  // entries — the regression these fixtures exist to catch.
  it("contains a case that distinguishes distinct-key counting from entry counting", () => {
    const discriminating = loadFixtures().superadmin_threshold.some(
      (c) =>
        c.expect === "error" &&
        c.min_valid_signatures > 0 &&
        c.signatures.length >= c.min_valid_signatures
    );
    expect(discriminating).toBe(true);
  });
});

// Step 5 — the per-group threshold. A different threshold from the SuperAdmin one above
// (per group, not tenant-wide) but the same counting rule, so it is gated from the same
// file: the four SDKs cannot drift on one without drifting on the other.
describe("signed fixtures: group threshold", () => {
  it("agrees with every recorded outcome", () => {
    const fixtures = loadFixtures();

    for (const c of fixtures.group_threshold) {
      const rules = {
        users: c.users.map((u) => ({
          id: u.user_id,
          publicKeyPem: u.public_key_pem,
          roles: ["USER"],
        })),
        groups: [{ id: c.group_id, userIds: c.group_user_ids }],
      } as unknown as DecodedRulesContainer;

      const signatures = c.signatures.map(
        (s) =>
          ({
            hashes: s.hashes,
            userSignature: { userId: s.user_id, signature: s.signature },
          }) as WhitelistSignatureEntry
      );

      const precomputed = new Map<number, Buffer>();
      signatures.forEach((sig, idx) => {
        precomputed.set(idx, Buffer.from(JSON.stringify(sig.hashes), "utf-8"));
      });

      const verifier = new WhitelistedAddressVerifier({
        superAdminKeysPem: [encodePublicKeyPem(
          crypto.generateKeyPairSync("ec", { namedCurve: "prime256v1" }).publicKey
        )],
        minValidSignatures: 1,
      });

      // verifyGroupThreshold is private; the walk is what is under test.
      const failure = (
        verifier as unknown as {
          verifyGroupThreshold: (
            t: GroupThreshold,
            r: DecodedRulesContainer,
            s: WhitelistSignatureEntry[],
            h: string,
            p: Map<number, Buffer>,
            k: Map<string, crypto.KeyObject>
          ) => string | null;
        }
      ).verifyGroupThreshold(
        {
          groupId: c.group_id,
          minimumSignatures: c.minimum_signatures,
          threshold: 0,
        } as GroupThreshold,
        rules,
        signatures,
        c.metadata_hash,
        precomputed,
        new Map()
      );

      // Assert as a tuple so a failing loop iteration names itself.
      expect([c.description, failure === null]).toEqual([
        c.description,
        c.expect === "ok",
      ]);
    }
  });

  // Same non-vacuity guard as the SuperAdmin section, applied per group.
  it("contains a case that distinguishes distinct-signer counting from entry counting", () => {
    const discriminating = loadFixtures().group_threshold.some((c) => {
      if (c.expect !== "error" || c.minimum_signatures <= 0) {
        return false;
      }
      const members = new Set(c.group_user_ids);
      const inGroup = c.signatures.filter((s) => members.has(s.user_id)).length;
      return inGroup >= c.minimum_signatures;
    });
    expect(discriminating).toBe(true);
  });
});

File: taurus-protect-sdk-go/pkg/protect/helper/signed_fixtures_test.go
package helper

import (
	"crypto/ecdsa"
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/crypto"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// Cross-SDK signed fixtures for verification step 2 — the SuperAdmin threshold.
//
// The behaviour vectors cover everything expressible without key material; this file
// covers what needs real signatures. Until it existed, the repo's own CLAUDE.md
// recorded that NO automated cross-SDK gate covered the distinct-key rule, and the
// five cases below lived as five hand-maintained copies in four suites.
//
// The fixture carries PUBLIC keys and signatures only — the gate verifies, it never
// signs — so there is no private key material in the repo.
const signedFixturesRelPath = "../../../../scripts/resources/verification-signed-fixtures.json"

type signedFixtureVector struct {
	Description string `json:"description"`
	Signatures  []struct {
		UserID    string `json:"user_id"`
		Signature string `json:"signature"`
	} `json:"signatures"`
	SuperAdminKeysPEM  []string `json:"super_admin_keys_pem"`
	MinValidSignatures int      `json:"min_valid_signatures"`
	Expect             string   `json:"expect"`
}

// Step-5 vector. The container's users and group membership are given, so the loader
// builds the same DecodedRulesContainer the verified container would have produced.
type groupFixtureVector struct {
	Description string `json:"description"`
	Users       []struct {
		UserID       string `json:"user_id"`
		PublicKeyPEM string `json:"public_key_pem"`
	} `json:"users"`
	Signatures []struct {
		UserID    string   `json:"user_id"`
		Signature string   `json:"signature"`
		Hashes    []string `json:"hashes"`
	} `json:"signatures"`
	MetadataHash      string   `json:"metadata_hash"`
	GroupID           string   `json:"group_id"`
	GroupUserIDs      []string `json:"group_user_ids"`
	MinimumSignatures int      `json:"minimum_signatures"`
	Expect            string   `json:"expect"`
}

type signedFixtures struct {
	RulesContainerBase64 string `json:"rules_container_base64"`
	Counts               struct {
		SuperAdminThreshold int `json:"superadmin_threshold"`
		GroupThreshold      int `json:"group_threshold"`
	} `json:"counts"`
	SuperAdminThreshold []signedFixtureVector `json:"superadmin_threshold"`
	GroupThreshold      []groupFixtureVector  `json:"group_threshold"`
}

func loadSignedFixtures(t *testing.T) signedFixtures {
	t.Helper()

	path := filepath.Clean(signedFixturesRelPath)
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("cannot read shared signed fixtures %s: %v", path, err)
	}

	var f signedFixtures
	if err := json.Unmarshal(raw, &f); err != nil {
		t.Fatalf("cannot parse %s: %v", path, err)
	}
	if len(f.SuperAdminThreshold) != f.Counts.SuperAdminThreshold {
		t.Fatalf("superadmin_threshold: got %d vectors, file declares %d",
			len(f.SuperAdminThreshold), f.Counts.SuperAdminThreshold)
	}
	if len(f.GroupThreshold) != f.Counts.GroupThreshold {
		t.Fatalf("group_threshold: got %d vectors, file declares %d",
			len(f.GroupThreshold), f.Counts.GroupThreshold)
	}
	if f.RulesContainerBase64 == "" {
		t.Fatal("fixture carries no rules container to verify against")
	}
	return f
}

func TestSignedFixtures_SuperAdminThreshold(t *testing.T) {
	fixtures := loadSignedFixtures(t)

	container, err := base64.StdEncoding.DecodeString(fixtures.RulesContainerBase64)
	if err != nil {
		t.Fatalf("cannot decode the fixture rules container: %v", err)
	}

	for _, tc := range fixtures.SuperAdminThreshold {
		t.Run(tc.Description, func(t *testing.T) {
			keys := make([]*ecdsa.PublicKey, 0, len(tc.SuperAdminKeysPEM))
			for i, keyPEM := range tc.SuperAdminKeysPEM {
				key, err := crypto.DecodePublicKeyPEM(keyPEM)
				if err != nil {
					t.Fatalf("cannot decode fixture key %d: %v", i, err)
				}
				keys = append(keys, key)
			}

			signatures := make([]*model.RuleUserSignature, 0, len(tc.Signatures))
			for _, s := range tc.Signatures {
				signatures = append(signatures, &model.RuleUserSignature{
					UserID:    s.UserID,
					Signature: s.Signature,
				})
			}

			err := VerifyGovernanceRulesSignatures(container, signatures, keys, tc.MinValidSignatures)

			if tc.Expect == "error" {
				if err == nil {
					t.Fatal("expected verification to fail, it passed")
				}
				return
			}
			if err != nil {
				t.Fatalf("expected verification to pass, got: %v", err)
			}
		})
	}
}

// The fixture must actually exercise the distinct-key rule: without a case where the
// signature COUNT meets the threshold but the distinct-key count does not, the whole
// file would pass against an implementation that counts entries.
func TestSignedFixtures_CoverTheDistinctKeyRule(t *testing.T) {
	fixtures := loadSignedFixtures(t)

	found := false
	for _, tc := range fixtures.SuperAdminThreshold {
		if tc.Expect == "error" && len(tc.Signatures) >= tc.MinValidSignatures && tc.MinValidSignatures > 0 {
			found = true
			break
		}
	}
	if !found {
		t.Error("no vector where entry count meets the threshold but distinct keys do not; " +
			"an implementation counting entries would pass this file")
	}
}

// Step 5 — the per-group threshold. A different threshold from the SuperAdmin one above
// (per group, not tenant-wide) but the same counting rule, so it is gated from the same
// file: the four SDKs cannot drift on one without drifting on the other.
func TestSignedFixtures_GroupThreshold(t *testing.T) {
	fixtures := loadSignedFixtures(t)

	for _, tc := range fixtures.GroupThreshold {
		t.Run(tc.Description, func(t *testing.T) {
			users := make([]*model.RuleUser, 0, len(tc.Users))
			for i, u := range tc.Users {
				key, err := crypto.DecodePublicKeyPEM(u.PublicKeyPEM)
				if err != nil {
					t.Fatalf("cannot decode fixture key for user %d: %v", i, err)
				}
				users = append(users, &model.RuleUser{ID: u.UserID, PublicKey: key})
			}

			container := &model.DecodedRulesContainer{
				Users:  users,
				Groups: []*model.RuleGroup{{ID: tc.GroupID, UserIDs: tc.GroupUserIDs}},
			}

			signatures := make([]model.WhitelistSignature, 0, len(tc.Signatures))
			for _, s := range tc.Signatures {
				signatures = append(signatures, model.WhitelistSignature{
					Hashes: s.Hashes,
					UserSignature: &model.WhitelistUserSignature{
						UserID:    s.UserID,
						Signature: s.Signature,
					},
				})
			}

			gt := &model.GroupThreshold{GroupID: tc.GroupID, MinimumSignatures: tc.MinimumSignatures}
			err := verifyGroupThreshold(gt, container, signatures, tc.MetadataHash,
				precomputeHashesJSON(signatures))

			if tc.Expect == "error" {
				if err == nil {
					t.Fatal("expected the threshold to fail, it passed")
				}
				return
			}
			if err != nil {
				t.Fatalf("expected the threshold to pass, got: %v", err)
			}
		})
	}
}

// Same non-vacuity guard as the SuperAdmin section: without a vector where the entry
// count meets the threshold but the distinct-signer count does not, the group section
// would pass against an implementation that counts entries.
func TestSignedFixtures_GroupSectionCoversTheDistinctSignerRule(t *testing.T) {
	fixtures := loadSignedFixtures(t)

	found := false
	for _, tc := range fixtures.GroupThreshold {
		if tc.Expect != "error" || tc.MinimumSignatures <= 0 {
			continue
		}
		inGroup := 0
		members := make(map[string]bool, len(tc.GroupUserIDs))
		for _, id := range tc.GroupUserIDs {
			members[id] = true
		}
		for _, s := range tc.Signatures {
			if members[s.UserID] {
				inGroup++
			}
		}
		if inGroup >= tc.MinimumSignatures {
			found = true
			break
		}
	}
	if !found {
		t.Error("no group vector where in-group entry count meets the threshold but " +
			"distinct signers do not; an implementation counting entries would pass")
	}
}
